frontend/english-keyset.ts

```diff
@@ -0,0 +1,339 @@
+                let onMouseDownFunction = this.getMouseDownFunction(key_str);
+                let onMouseUpFunction = this.getMouseUpFunction(key_str);
+
+                keys.push(
+                    new Key(
+                        "",
+                        key_str,
+                        "",
+                        1,
+                        is_special,
+                        span,
+                        onDownFunction,
+                        onUpFunction,
+                        onMouseDownFunction,
+                        onMouseUpFunction,
+                    ),
+                );
+            });
+            return keys;
+        });
+    }
+
+    private getKeyDownFunction(key_str: string) {
+        return (e: KeyboardEvent) => { console.log(key_str + " Key Down") };
+    }
+
+    private getKeyUpFunction(key_str: string) {
+        return (e: KeyboardEvent) => { console.log(key_str + " Key Up") };
+    }
+
+    private getMouseDownFunction(key_str: string) {
+
+        let customMouseDownFn = () => {
+
+            if (key_str.toLowerCase() in ["shift", "ctrl", "alt"]) {
+                let curKeyState = Boolean(this.containingKeyboard.states[key_str.toLowerCase()]);
+                this.containingKeyboard.states[key_str.toLowerCase()] = !curKeyState;
+            }
+            else {
+
+                if (this.containingKeyboard.targetElement instanceof HTMLInputElement ||
+                    this.containingKeyboard.targetElement instanceof HTMLTextAreaElement) {
+
+                    let curValue = this.containingKeyboard.targetElement.value;
+                    let modifiedValue = getModifiedValueOnKeyPress(curValue, key_str, this.containingKeyboard.states);
+                    this.containingKeyboard.targetElement.value = modifiedValue;
+
+                }
+                else if (typeof (this.containingKeyboard.targetElement) === "string") {
+
+                    let curValue = this.containingKeyboard.targetElement;
+                    let modifiedValue = getModifiedValueOnKeyPress(curValue, key_str, this.containingKeyboard.states);
+                    this.containingKeyboard.targetElement = modifiedValue;
+                }
+            }
+
+            if (this.containingKeyboard.targetElement instanceof HTMLElement) {
+                this.containingKeyboard.targetElement.dispatchEvent(new Event('input'));  // notify value update
+                this.containingKeyboard.targetElement.dispatchEvent(new KeyboardEvent('keydown', {
+                    key: key_str,
+                    code: key_str,
+                    ctrlKey: Boolean(this.containingKeyboard.states['ctrl']),
+                    shiftKey: Boolean(this.containingKeyboard.states['shift']),
+                    altKey: Boolean(this.containingKeyboard.states['alt']),
+                }));
+            }
+        };
+
+        return () => {
+            customMouseDownFn();
+        };
```

Comment:
interesting

Reply:
Ah,

It's like this because initially the anonymous function was accepting the mouse event param incase the customFunction needed it, but I removed it later. I'll change it